### DISCUSSION TITLE
Fixing the size of cornea

### DISCUSSION THREAD

**Discussion by Mayank4884** [openPupil/Open-PupilEXT#49] [Ideas]
Since there is an expected behaviour that when we use a single camera and the pixel-size of the circular object changes with different distances to the camera.. Why don't you fix the size of the cornea.. And then measure.. As we know size of cornea is constant in all individuals.. So even if there is change in distance.. The measurements won't change right.. 
Please add this feature in the app.. To fix the size of cornea.. It will be very easy for detecting the size of the pupil for externally loaded images
Thankyou

**Comment by BZandi** [comment]
When recording with a single camera, the pupil size varies with different working distances. This is an expected behaviour, meaning there is nothing to fix as this is no error. However, in such experiments (single camera), it is highly recommended to use a chin rest to keep the working distance steady. If you like to use PupilEXT in experimental conditions where the working distance varies dynamically, you need to use two cameras (stereo camera system). This is supported in PupilEXT.

However, when using a single camera, It is also possible to attach a circular reference object below the subject's eye. In this way, you can fix the working-size-to-pixel ratio also with one camera without having to use a stereo camera system, as you always have a reference object in the image. This is also possible when using PupilEXT.

Please keep in mind that is an open-source project, which is mainly maintained in my spare time in the hope that PupilEXT can support the research community. There is no warranty, without even the implied warranty of fitness for a particular purpose.

If you are highly interested in new features, you are free to implement those in PupilEXT yourself. The source code is open-source. We also added a new community branch (https://github.com/openPupil/Open-PupilEXT/tree/Experimental-Community-Version), allowing PupilEXT users to add new features via pull request.

**Reply by Mayank4884** [reply]
I asking to fix the size of cornea not pupil.. The heading was typo

**Comment by kheki4** [comment]
Hi @Mayank4884 , your suggestion is interesting. But actually, we are not yet at the stage where we design our own pupil detection algorithm. Probably one or two years from now, we will get there in capacity.

Anyway, I would like to add my personal opinion on this idea (in case you, or someone reading this, is trying to develop their own algorithm):
- Even the size of the cornea can be different between individuals, so its size could not be used as an absolute value among all subjects.
- Cornea can be easily occluded by eyelids, thus only the left/right quarters of it could be used reliably in an algorithm.
- Selecting the pupil by brightness thresholding is feasible.. selecting corneal light reflection(s) too.. but the cornea can be different based one one's eye colour (even if we are talking about NIR light), so thresholding would need to rely on heuristics or used defined values in my opinion.
- It is way more stable to apply a reference object, like a sticker on the subject's face. In pupilEXT, using a single camera with a chin rest and you ensure a constant face-camera distance, in image recordings you can do this, then manually fix the measured pupil size once the csv file is generated. Reference object is also used by e.g. SR Research EyeLink branded eye trackers, they can precisely track it programmatically, and the output values are already fixed by that. 
 
Otherwise I think you could rely on tracking both eyes, and (heavily hoping that both eyes are well detected on their own) use intercanthal distance as a possible constant distance - but even if their distance is known as constant, the actual physical distance should be told by the user, because canthi can be very different between individuals. 

Aside from these, you could also use landmark detection, like detecting the tip of the nose, or other parts like eyebrows of the face, and use e.g. a triangle of them to make constants. But they are also computationally expensive, and need to be combined with proper tracking (= involving the detection outcome of the previous frame(s) in the detection using the present frame) and some heuristics (e.g. the nose tip will always be on the bottom half of the mid-perpendicular between the canthi). So optimally solving this can really go deep in the rabbit hole :)
